File: frontend/types/workflow.ts
export type WorkflowStatus =
  | 'PENDING'
  | 'QUEUED'
  | 'RUNNING'
  | 'WAITING_APPROVAL'
  | 'RETRYING'
  | 'COMPLETED'
  | 'FAILED'
  | 'CANCELLED';

export type StepStatus =
  | 'PENDING'
  | 'RUNNING'
  | 'WAITING_APPROVAL'
  | 'COMPLETED'
  | 'FAILED'
  | 'SKIPPED';

export interface WorkflowStep {
  id: string;
  workflow_id: string;
  step_key: string;
  agent_type: string;
  dependency_ids: string[];
  input_data: Record<string, any>;
  output_data: Record<string, any>;
  status: StepStatus;
  retry_count: number;
  error_message?: string;
  execution_time: number;
  created_at: string;
}

export interface Workflow {
  id: string;
  name: string;
  goal: string;
  status: WorkflowStatus;
  user_id?: string;
  execution_time: number;
  meta_data: Record<string, any>;
  created_at: string;
  updated_at: string;
  steps: WorkflowStep[];
}

export interface WorkflowTemplate {
  id: string;
  title: string;
  description: string;
  category: string;
  goal: string;
  agent_chain: string[];
  icon: string;
  is_preset: boolean;
  created_at: string;
}

export interface WorkflowCreatePayload {
  name?: string;
  goal: string;
  template_id?: string;
  meta_data?: Record<string, any>;
}

File: backend/app/schemas/workflow.py
from typing import List, Dict, Any, Optional
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from app.models.workflow import WorkflowStatus
from app.models.workflow_step import StepStatus

class WorkflowCreate(BaseModel):
    name: Optional[str] = None
    goal: str = Field(..., description="High-level objective for the multi-agent system")
    template_id: Optional[str] = None
    meta_data: Optional[Dict[str, Any]] = Field(default_factory=dict)

class WorkflowStepResponse(BaseModel):
    id: str
    workflow_id: str
    step_key: str
    agent_type: str
    dependency_ids: List[str]
    input_data: Dict[str, Any]
    output_data: Dict[str, Any]
    status: StepStatus
    retry_count: int
    error_message: Optional[str]
    execution_time: float
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class WorkflowResponse(BaseModel):
    id: str
    name: str
    goal: str
    status: WorkflowStatus
    user_id: Optional[str]
    execution_time: float
    meta_data: Dict[str, Any]
    created_at: datetime
    updated_at: datetime
    steps: List[WorkflowStepResponse] = []

    model_config = ConfigDict(from_attributes=True)

class HumanApprovalRequest(BaseModel):
    approved: bool
    feedback: Optional[str] = None

class WorkflowFilterParams(BaseModel):
    status: Optional[WorkflowStatus] = None
    search: Optional[str] = None
    limit: int = 50
    offset: int = 0
